File: Xholon/src/org/primordion/user/app/climatechange/carboncycle03/Upwelling.java
package org.primordion.user.app.climatechange.carboncycle03;

import org.primordion.xholon.base.IXholon;

public class Upwelling extends Xhcarboncycle03 {
	
	private double rateFactor = 0.0;
	
	private IXholon surfaceOcean = null;
	private IXholon deepOcean = null;
	
	public void postConfigure() {
		surfaceOcean = ((Flows)getParentNode()).getSurfaceOcean();
		deepOcean = ((Flows)getParentNode()).getDeepOcean();
		rateFactor = this.getVal() / deepOcean.getVal();
		super.postConfigure();
	}
	
	public void act() {
		double rate = this.getVal();
		surfaceOcean.incVal(rate);
		deepOcean.decVal(rate);
		super.act();
	}
	
	public void postAct() {
		this.setVal(deepOcean.getVal() * rateFactor);
		super.postAct();
	}

}

File: Xholon/war/config/climatechange/carboncycle03/script/javascript/Upwelling.js
<Upwelling val_double="101.0" implName="org.primordion.user.app.climatechange.carboncycle03.Flow_javascript"><![CDATA[
  var behaviorObject = new Object();
  
  behaviorObject.rateFactor = 0.0;
  behaviorObject.surfaceOcean = null;
  behaviorObject.deepOcean = null;
  
  behaviorObject.postConfigure = function() {
    surfaceOcean = contextNodeKey.getParentNode().getSurfaceOcean();
    deepOcean = contextNodeKey.getParentNode().getDeepOcean();
    rateFactor = contextNodeKey.getVal() / deepOcean.getVal();
  }
  
  behaviorObject.act = function() {
    var rate = contextNodeKey.getVal();
    surfaceOcean.incVal(rate);
    deepOcean.decVal(rate);
  }
  
  behaviorObject.postAct = function() {
    contextNodeKey.setVal(deepOcean.getVal() * rateFactor);
  }

]]></Upwelling>